rules/alerting.go

```diff
@@ -316,20 +315,58 @@ func (r *AlertingRule) Restored() bool {
 	return r.restored.Load()
 }
 
-func (r *AlertingRule) SetNoDependentRules(noDependentRules bool) {
-	r.noDependentRules.Store(noDependentRules)
+func (r *AlertingRule) SetDependentRules(dependents []Rule) {
+	r.dependenciesMutex.Lock()
+	defer r.dependenciesMutex.Unlock()
+
+	r.dependentRules = make(map[string]struct{})
+	for _, dependent := range dependents {
+		r.dependentRules[dependent.Name()] = struct{}{}
+	}
 }
 
 func (r *AlertingRule) NoDependentRules() bool {
-	return r.noDependentRules.Load()
+	r.dependenciesMutex.RLock()
+	defer r.dependenciesMutex.RUnlock()
+
+	if r.dependentRules == nil {
+		return false // We don't know if there are dependent rules.
+	}
+
+	return len(r.dependentRules) == 0
+}
+
+func (r *AlertingRule) DependentRules() map[string]struct{} {
+	r.dependenciesMutex.RLock()
+	defer r.dependenciesMutex.RUnlock()
+	return r.dependentRules
 }
 
-func (r *AlertingRule) SetNoDependencyRules(noDependencyRules bool) {
-	r.noDependencyRules.Store(noDependencyRules)
+func (r *AlertingRule) SetDependencyRules(dependencies []Rule) {
+	r.dependenciesMutex.Lock()
+	defer r.dependenciesMutex.Unlock()
+
+	r.dependencyRules = make(map[string]struct{})
+	for _, dependency := range dependencies {
+		r.dependencyRules[dependency.Name()] = struct{}{}
```

Comment:
You can technically have two rules with the same name in the same group - how does this affect this implementation? Shouldn't you also include the index as part of the key?

Reply:
Modified to store a slice of Rule instead: https://github.com/prometheus/prometheus/pull/15689/commits/9e8a5fb0a4ba73a5e9212a7237695419b518f5f5